merge_cubes.json

```diff
@@ -1,7 +1,7 @@
ng dimension has the same name, type, reference system and resolution in both dimensions, but can have different labels. One of the dimensions can have different labels, for all other dimensions the labels must be equal. If data overlaps, the parameter `overlap_resolver` must be specified to resolve the overlap.\n\n**Examples for merging two data cubes:**\n\n1. Data cubes with the dimensions (`x`, `y`, `t`, `bands`) have the same dimension labels in `x`, `y` and `t`, but the labels for the dimension `bands` are `B1` and `B2` for the first cube and `B3` and `B4`. An overlap resolver is *not needed*. The merged data cube has the dimensions `x`, `y`, `t` and `bands` and the dimension `bands` has four dimension labels: `B1`, `B2`, `B3`, `B4`.\n2. Data cubes with the dimensions (`x`, `y`, `t`, `bands`) have the same dimension labels in `x`, `y` and `t`, but the labels for the dimension `bands` are `B1` and `B2` for the first data cube and `B2` and `B3` for the second. An overlap resolver is *required* to resolve overlap in band `B2`. The merged data cube has the dimensions `x`, `y`, `t` and `bands` and the dimension `bands` has three dimension labels: `B1`, `B2`, `B3`.\n3. Data cubes with the dimensions (`x`, `y`, `t`) have the same dimension labels in `x`, `y` and `t`. There are two options:\n   1. Keep the overlapping values separately in the merged data cube: An overlap resolver is *not needed*, but for each data cube you need to add a new dimension using ``add_dimension()``. The new dimensions must be equal, except that the labels for the new dimensions must differ by name. The merged data cube has the same dimensions and labels as the original data cubes, plus the dimension added with ``add_dimension()``, which has the two dimension labels after the merge.\n   2. Combine the overlapping values into a single value: An overlap resolver is *required* to resolve the overlap for all values. The merged data cube has the same dimensions and labels as the original data cubes, but all values have been processed by the overlap resolver.\n4. A data cube with dimensions (`x`, `y`, `t` / `bands`) or (`x`, `y`, `t`, `bands`) and another data cube with dimensions (`x`, `y`) have the same dimension labels in `x` and `y`. Merging them will join dimensions `x` and `y`, so the lower dimension cube is merged with each time step and band available in the higher dimensional cube. This can for instance be used to apply a digital elevation model to a spatio-temporal data cube. An overlap resolver is *required* to resolve the overlap for all pixels.\n\nAfter the merge, the dimensions with a natural/inherent label order (with a reference system this is each spatial and temporal dimensions) still have all dimension labels sorted. For other dimensions where there is no inherent order, including bands, the dimension labels keep the order in which they are present in the original data cubes and the dimension labels of `cube2` are appended to the dimension labels of `cube1`.",
```

Comment:
What happens if I try to merge a raster cube and a vector cube? Is this an operation that is even defined? Sorry if this has an obvious answer, but purely from this description I wouldn't know this!

Reply:
I haven't really finalized merge_cubes yet, this is still WIP. What would you expect?

Reply:
Hmm. My first naive thought was that you could either rasterise the vector data or vectorise the raster data, but both of that gets weird really quickly and even weirder when considering how overlap resolvers would work. 

Also at least for EODC I don't see how we'd implement such a merge_cubes process with our current xarray/geopandas stack (maybe something helpful will come out of [this thread](https://discourse.pangeo.io/t/vector-data-cubes/2904) on the pangeo discourse). So would be in favour of this being undefined behaviour until it becomes clearer what that behaviour should even be and we have a usecase that requires it!

Reply:
Ah! I think merging a vector and a raster data cube is not allowed in merge_cubes because they're not "compatible". I should mention this explicitly.

Another thought I had today is how merging two vector cubes would work. Issues could arise for an equality check based on the geometry (floating point imprecision) and/or the id (e.g. the same id in two datasets or applying a buffer on a vector cube without changing the ID).

Hmm...

Reply:
> Ah! I think merging a vector and a raster data cube is not allowed in merge_cubes because they're not "compatible". I should mention this explicitly.

Ah, I'm glad, that simplifies things :D

> There are three options I assume:
> 
> geometry: floating point imprecision might arise
> id: vector cube A with unique IDs -> apply buffer (and results in vector cube B), IDs don't change (I assume?) -> merge them, which would lead to a conflict / unpredictable behavior. So either we need to change the IDs when the geometries change or ...?
> Rely on whatever the software being used considers to be equal (I assume that's aligned with SF). I assume that's our best bet, but may lead to different results across backends?
> Thoughts?

- geometry: What about using an `almost_equal` scheme (akin to `GeoSeries.geom_almost_equals`) with an openeo-wide precision setting? I can't think of a case where that wouldn't be appropriate off the top of my head?
- ID: Where were you thinking to get these from in the first place? In shapely 2.0 geometries will become hashable, based on their WKB representation - could that be an option? 

These we'd probably just decide on a convention for, but two other factors that could influence equality:
- row/column order 
- CRS